### PR TITLE
Enable the use of different number formats e.g. 2G, 2048M or 2147483648

### PR DESCRIPTION
**Pull request by ag-wood**
Addresses issue #879 while preserving current functionality (i.e. no restriction on number format)

### CODE REVIEW

File: system/zfs.py [hunk 1]
```diff
@@ -237,6 +237,36 @@
 
 import os
 
+def to_bytes(value):
```
Comment (bcoca) [inline system/zfs.py]: these functions would belong in a lib/ansible/module_utils/basic.py (see existing pretty_bytes)

**Comment by ag-wood** [comment]
Thanks @bcoca for the suggestion re: module_utils.   Do I just submit a separate pull request to ansible/ansible?  Also, not being terribly experienced with how things get done, am curious about how the separate requests would get merged as this creates a dependency for the inclusion of functionality in the zfs module.  Sorry if these are dumb questions.

**Comment by Yannig** [comment]
Hi @ag-wood,

I have already submited a pull request matching this needs at the following location: https://github.com/ansible/ansible/pull/12074

I'm using an ansible filter to convert size strings into bytes.

What do you think about @bcoca?

**Comment by ag-wood** [comment]
Does that help in the context that I need it?  Does a filter get used like other functions within a module ?i.e. I wanted to use the function for testing a value passed into the zfs module.

**Comment by ag-wood** [comment]
Based on @bcoca 's recommendation, I was planning to submit the following two functions:

```
def to_bytes(self, size):
    if isnumeric(size):
        return size
    suffixes = ['B', 'K', 'M', 'G', 'T', 'P', 'E', 'Z', 'Y']
    index = next(i for i, d in enumerate(size) if not d.isdigit())
    if index > 0:
        value = size[:index]
        suffix = (size[index:][0]).upper()
        if suffix in suffixes:
            power = suffixes.index(suffix)
            return int(float(value) * (1024 ** power))
    return None

def isnumeric(self, value):
    try:
        v = float(value)
        return True
    except:
        return False
```

**Comment by bcoca** [comment]
i had forgotten about the filters, they can be used as input to the modules, not from the modules themselves. Having the fitlers I don't think the modules themselves need to handle these things.

also the is_numeric function is just check for floats, that can be handled by setting the type in the module spec to float.

**Comment by ag-wood** [comment]
The is_numeric and to_bytes function are purely to manage the string values that can be passed (i.e. to check whether the string is numeric as opposed to any of the other allowable string values i.e. 2G, 2048M etc.)  I don't want to restrict the input to a numeric value as this hamstrings the module and requires the use of very long numbers that are easy to get mixed up.

Essentially, I was trying to add some flexibility to satisfy issue #879 without making it more difficult to use, which I think would be the outcome if we force numeric inputs for the size values.

I'm also having trouble seeing how the human_bytes filter can be used within the module for post processing inputs in the manner described above.

**Comment by bcoca** [comment]
im saying that the 'general interface' will allow for this while not needing to add this complexity to modules which can take a stricter more specific input:

`size: "{{20G|to_bytes}}"`

having filters that do this removes the need for modules to handle it

**Comment by ag-wood** [comment]
It still seems to be less straight forward than handling it within the module.  I get that an Ansible savvy operator might fall into the pattern you are suggesting, but I don't see that it fits with the 'Ansible is simple' mantra.

I'm willing to concede defeat and bow to the greater wisdom - so the summary then is that you prefer to change the interface (force numerics for quota, maybe volsize...) which may break existing playbooks.  I'm not prepared to make that sort of call as I'm not sure how backward compatibility should work.

**Comment by bcoca** [comment]
or let me completely reverse my position, create a 'bytesize' module type that does this as part of the module spec

**Comment by ag-wood** [comment]
So, do you mean double up on each parameter that can be represented as a string or numeric?

**Comment by bcoca** [comment]
no, just have a type='bytes', which accepts strings and returns a normalized 'byte size'

**Comment by Yannig** [comment]
@bcoca: It's a good idea.

I have start working on a little patch:

``` diff
diff --git a/lib/ansible/module_utils/basic.py b/lib/ansible/module_utils/basic.py
index 60fe323..cffad59 100644
--- a/lib/ansible/module_utils/basic.py
+++ b/lib/ansible/module_utils/basic.py
@@ -470,6 +470,7 @@ class AnsibleModule(object):
                 'int': self._check_type_int,
                 'float': self._check_type_float,
                 'path': self._check_type_path,
+                'bytes': self._check_type_bytes,
             }
         if not bypass_checks:
             self._check_required_arguments()
@@ -1210,6 +1211,8 @@ class AnsibleModule(object):
         value = self._check_type_str(value)
         return os.path.expanduser(os.path.expandvars(value))

+    def _check_type_bytes(self, value):
+        return self.string_to_bytes(value)

     def _check_argument_types(self):
         ''' ensure all arguments have the requested type '''
@@ -1806,22 +1809,47 @@ class AnsibleModule(object):
         fh.write(str)
         fh.close()

+    size_ranges = (
+        (1<<80L, 'YB'),
+        (1<<70L, 'ZB'),
+        (1<<60L, 'EB'),
+        (1<<50L, 'PB'),
+        (1<<40L, 'TB'),
+        (1<<30L, 'GB'),
+        (1<<20L, 'MB'),
+        (1<<10L, 'KB'),
+        (1, 'Bytes')
+    )
     def pretty_bytes(self,size):
-        ranges = (
-                (1<<70, 'ZB'),
-                (1<<60, 'EB'),
-                (1<<50, 'PB'),
-                (1<<40, 'TB'),
-                (1<<30, 'GB'),
-                (1<<20, 'MB'),
-                (1<<10, 'KB'),
-                (1, 'Bytes')
-            )
-        for limit, suffix in ranges:
+        for limit, suffix in size_ranges:
             if size >= limit:
                 break
         return '%.2f %s' % (float(size)/ limit, suffix)

+    def string_to_bytes(size):
+        ''' Convert size in string format into bytes (ex: '2K' => 2048) '''
+        m = re.search('^\s*(\d*\.?\d*)\s*([A-Z])?', size, flags=re.IGNORECASE)
+        if m is None:
+            raise ValueError("string_to_bytes() can't interpret following string: %s" % str(size))
+        try:
+            num = float(m.group(1))
+        except:
+            raise ValueError("string_to_bytes() can't interpret following number: %s (original input string: %s)" % (m.group(1), size))
+
+        unit = m.group(2)
+        if unit is None:
+            ''' No unit given, returning raw number '''
+            return int(round(num))
+
+        unit = unit[0].upper() + 'B'
+
+        for limit, suffix in size_ranges:
+            if unit == suffix:
+                break
+        if unit != 'Bytes' and limit == 1:
+            raise ValueError("string_to_bytes() can't interpret following string: %s" % size)
+        return int(round(num * limit))
+
     #
     # Backwards compat
     #
```

I was starting to port lvol to this new type and find that size can be expressed in percent (50%).

@bcoca: what do you think about? Do we need to create a new type like bytes_or_percent?

**Comment by robynbergeron** [comment]
Hi @ag-wood (thanks for the PR!), @Yannig (thanks for the feedback and the other PR), @bcoca -- thanks for looking at this thus far. 

Also pinging @johanwiren (zfs module maintainer) to let you know about this PR. Johan - it looks like this may be redone as a wider PR to ansible/ansible -- but just putting this on your radar just in case, and for feedback anyway :)

Thanks, everyone! :)

**Comment by Yannig** [comment]
I have done some rework against my PR (https://github.com/ansible/ansible/pull/12074) in order to be able to call the same function from AnsibleModule and string_to_bytes from module_utils.

@bcoca, @ag-wood: what do you think about?

**Comment by gregdek** [comment]
Thanks @ag-wood for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

**Comment by gregdek** [comment]
@ag-wood A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes.

**Comment by ag-wood** [comment]
Closing this as I don't have the energy to convince anyone that my approach was worth implementing.

**Pull request closed without merging**